profiler/src/ProfilerEngine/Datadog.Profiler.Native/LibddprofExporter.cpp

```diff
@@ -312,30 +324,56 @@ void LibddprofExporter::Add(Sample const& sample)
     ffiSample.values = {values.data(), values.size()};
 
     ddprof_ffi_Profile_add(profile, ffiSample);
-    profileInfo.samplesCount++;
+    profileInfoScope.profileInfo.samplesCount++;
 }
 
 bool LibddprofExporter::Export()
 {
     bool exported = false;
 
     int32_t idx = 0;
-    for (auto& [runtimeId, profileInfo] : _perAppInfo)
+
+    std::vector<std::string_view> keys;
+
+    {
+        std::lock_guard lock(_profileInfoLock);
+        for (const auto& [key, _] : _perAppInfo)
+        {
+            keys.push_back(key);
+        }
+    }
+
+    for (auto& runtimeId : keys)
     {
-        auto samplesCount = profileInfo.samplesCount;
+        ddprof_ffi_Profile* profile;
+        int32_t samplesCount;
+        int32_t exportsCount;
+
+        {
+            const auto scope = GetInfo(runtimeId);
+
+            // Get everything we need then release the lock
+            profile = scope.profileInfo.profile;
+            samplesCount = scope.profileInfo.samplesCount;
+
+            // Count is incremented BEFORE creating and sending the .pprof
+            // so that it will be possible to detect "missing" profiles
+            // in the back end
+            exportsCount = ++scope.profileInfo.exportsCount;
+
+            scope.profileInfo.profile = nullptr;
+            scope.profileInfo.samplesCount = 0;
+        }
+
         const auto& applicationInfo = _applicationStore->GetApplicationInfo(std::string(runtimeId));
 
-        if (samplesCount <= 0)
+        if (profile == nullptr || samplesCount == 0)
         {
             Log::Debug("The profiler for application ", applicationInfo.ServiceName, " (runtime id:", runtimeId, ") have empty profile. Nothing will be send.");
```

Comment:
```suggestion
            Log::Debug("The profiler for application ", applicationInfo.ServiceName, " (runtime id:", runtimeId, ") have empty profile. Nothing will be sent.");
```

Reply:
Done in https://github.com/DataDog/dd-trace-dotnet/pull/2990/commits/f6491935e793d28a61b87bdd18c6e5175fbce622